comms/discovery/server/server.go

```diff
@@ -473,30 +510,74 @@ func (s *ChatServer) validateCanChat(c echo.Context) error {
 		return c.JSON(400, "bad request: "+err.Error())
 	}
 	var receiverUserIds []int32
-	validatedPermissions := make(map[string]bool)
+	canChat := make(map[string]bool)
 	for _, encodedId := range params.ReceiverUserIDS {
 		decodedId, err := misc.DecodeHashId(encodedId)
 		if err != nil {
 			return c.JSON(400, "bad request: "+err.Error())
 		}
 		receiverUserIds = append(receiverUserIds, int32(decodedId))
 		// Initialize response map
-		validatedPermissions[encodedId] = true
+		canChat[encodedId] = true
 	}
 
-	// Validate permission for each <request sender, user> pair
-	permissions, err := queries.BulkGetChatPermissions(db.Conn, c.Request().Context(), receiverUserIds)
-	if err != nil {
-		if err == sql.ErrNoRows {
-			response := schema.CommsResponse{
-				Health: s.getHealthStatus(),
-				Data:   validatedPermissions,
+	// Validate request sender is not the blocker or blockee for each receiver
+	blocks, err := queries.BulkGetChatBlocksOrReceivedBlocks(db.Conn, c.Request().Context(), queries.BulkGetChatBlocksOrReceivedBlocksParams{
+		SenderUserID:    userId,
+		ReceiverUserIDs: receiverUserIds,
+	})
+	if err != nil && err != sql.ErrNoRows {
+		return err
+	}
+	if err != sql.ErrNoRows {
+		var encodedId string
+		var receiver int32
+		var blockedReceivers []int32
+		for _, block := range blocks {
+			if block.BlockerUserID == userId {
+				// Request sender blocked this receiver
+				receiver = block.BlockeeUserID
+			} else {
+				// Receiver blocked request sender
+				receiver = block.BlockerUserID
 			}
-			return c.JSON(200, response)
-		} else {
+			// Update response map
+			encodedId, err = misc.EncodeHashId(int(receiver))
+			if err != nil {
+				return err
+			}
+			canChat[encodedId] = false
+			blockedReceivers = append(blockedReceivers, receiver)
+		}
+
+		// Remove blockedReceivers from receiverUserIds.
+		// We know sender cannot chat with blockedReceivers - no need
+		// to perform further validations.
+		receiverUserIds = utils.Difference(receiverUserIds, blockedReceivers)
+	}
+
+	if len(receiverUserIds) > 0 {
+		// Validate permission for each <request sender, user> pair
+		permissions, err := queries.BulkGetChatPermissions(db.Conn, c.Request().Context(), receiverUserIds)
+		if err != nil && err != sql.ErrNoRows {
 			return err
 		}
+		if err != sql.ErrNoRows {
+			canChat, err = validatePermissions(c, permissions, userId, canChat)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	response := schema.CommsResponse{
+		Health: s.getHealthStatus(),
+		Data:   canChat,
 	}
+	return c.JSON(200, response)
+}
+
+func validatePermissions(c echo.Context, permissions []queries.ChatPermissionsRow, senderUserId int32, canChat map[string]bool) (map[string]bool, error) {
```

Comment:
Love that this got split out. Wondering if we should do more of this and separate out some of the logic from this one large file.

Reply:
yeah can do as part of the rest -> rpc refactor later